### PR TITLE
KAFKA-15045: (KIP-924 pt. 9) TaskAssignmentUtils implementation of optimizeRackAwareActiveTasks

### PR DESCRIPTION
**Pull request by apourchet**
This PR implements the rack aware optimization of active tasks that can be used by the assignors themselves. It takes in the full output of the assignment and tries to reorganize it so as to minimize traffic and non-overlap costs.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskInfo.java [hunk 1]
```diff
@@ -50,6 +50,12 @@ public interface TaskInfo {
      */
     Set<String> stateStoreNames();
 
+    /**
+     *
+     * @return The topic partitions in use by this task.
+     */
+    Set<TaskTopicPartition> topicPartitions();
```
Comment (apourchet) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskInfo.java:55]: This is temporary to get things to compile until we merge part 10 https://github.com/apache/kafka/pull/16034/files

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java [hunk 2]
```diff
@@ -486,12 +486,12 @@ public long optimizeStandbyTasks(final SortedMap<UUID, ClientState> clientStates
...
                         .collect(Collectors.toList());
 
-                    final Map<TaskId, UUID> taskClientMap = new HashMap<>();
                     final List<UUID> clients = Stream.of(clientList.get(i), clientList.get(j))
                         .sorted().collect(
                             Collectors.toList());
-                    final Map<UUID, Integer> originalAssignedTaskNumber = new HashMap<>();
 
+                    final Map<TaskId, UUID> taskClientMap = new HashMap<>();
+                    final Map<UUID, Integer> originalAssignedTaskNumber = new HashMap<>();
```
Comment (apourchet) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java:475]: I moved this lower down to show them clearly initialized empty right before the function call. It turns out there parameters are used as outputs, which is why I added the `AssignmentGraph` inner class to the `TaskAssignmentUtils`, in order to simplify the usage of the graph functions.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 3]
```diff
@@ -39,6 +91,8 @@ public static Map<ProcessId, KafkaStreamsAssignment> defaultStandbyTaskAssignmen
         final ApplicationState applicationState,
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments
     ) {
+        final long followupRebalanceDelay = applicationState.assignmentConfigs().probingRebalanceIntervalMs();
+        final Instant followupRebalanceDeadline = Instant.now().plus(followupRebalanceDelay, ChronoUnit.MILLIS);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Was this intentional? FWIW none of the `TaskAssignmentUtils` methods should ever schedule followup rebalances

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 4]
```diff
@@ -58,7 +112,85 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTask
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:104]: It might be worth including some of the javadocs from RackAwareTaskAssignor#optimizeActiveTasks, especially the ones that describe what the two relevant configs do/mean. I'll put something in the KIP and you can just copy/paste from there

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 5]
```diff
@@ -74,20 +209,240 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+        }
+
+        if (task.sourceTopicPartitions().isEmpty()) {
+            LOG.error("Task {} has no source TopicPartitions.", task.id());
+            return false;
+        }
+
+        if (hasStandby && task.changelogTopicPartitions().isEmpty()) {
+            LOG.error("Task {} has no changelog TopicPartitions.", task.id());
+            return false;
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: This should not be an error, we can remove the `hasStandby` parameter entirely.

For one thing, it shouldn't be an error to have configured an app to use standby tasks even if there's no state/changelog topics (Streams just wouldn't create those standby tasks). But for another, it's also possible to have a standby task with no changelog topics. Not all stateful tasks necessarily have to have changelogs (though most do), one can configure a state store with "logging disabled" which just means it has no changelog topic. Sometimes people play fast and lose with semantics

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 6]
```diff
@@ -74,20 +209,240 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+
+        for (final TopicPartition topicPartition : task.changelogTopicPartitions()) {
+            final Set<String> racks = racksByTopicPartition.getOrDefault(topicPartition, null);
+            if (racks == null || racks.isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+
+        if (task.sourceTopicPartitions().isEmpty()) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: It really doesn't make sense to include this check here, whether a task has source partitions has nothing to do with whether the rack info is valid. Also, it shouldn't be up to the user/assignor to verify the inputs to the assignor and check conditions such as "every task has at least one source topic partition".

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 7]
```diff
@@ -58,7 +112,88 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTask
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
     ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+        if (tasks.isEmpty()) {
+            return kafkaStreamsAssignments;
+        }
+
+        if (!hasValidRackInformation(applicationState)) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: It seems a little silly to have to run this exact check twice, which we are essentially doing by throwing an exception if it fails. For example the StickyTaskAssignor runs this check and if false, skips calling the rack-aware optimization. 

Imo, to best keep the current behavior where it simply skips applying this optimization if it doesn't have all of the rack id info it needs, we should just return early here instead of throwing an IllegalStateException, and not call it at all in the StickyTaskAssignor (or any of the assignors). The point is to move any rack-related logic out of the assignors and encapsulate it in these utility methods, so they can simply be called without knowing anything about how rack aware assignment works. Throwing an exception if you call this method without first having run some pre-check that users won't easily know about seems a bit cruel

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 8]
```diff
@@ -74,20 +209,240 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
+        final RackAwareGraphConstructor<UUID> graphConstructor = new MinTrafficGraphConstructor<>();
+        final AssignmentGraph assignmentGraph = buildTaskGraph(clientIds, clientRacks, taskIds, taskIdsByProcess, topicPartitionsByTaskId,
+            crossRackTrafficCost, nonOverlapCost, hasReplica, isStandby, graphConstructor);
+        return assignmentGraph.graph.totalCost();
+    }
+
+    private static AssignmentGraph buildTaskGraph(final List<UUID> clientIds,
+                                                 final Map<UUID, Optional<String>> clientRacks,
+                                                 final List<TaskId> taskIds,
+                                                 final Map<UUID, Set<TaskId>> taskIdsByProcess,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: We're passing in the wrong thing here. Ultimately this is used for the `hasAssignedTask` predicate by the graph constructors, which means "is this task assigned to this client in the current iteration of the assignment". However both callers of this method construct the `tasksIdsByProcess` map based on the _previous_ task assignment, not the _current_ one. 

We need to be careful in differentiating between previous assignment and current assignment. How are we actually keeping track of and/or updating the current assignment through these rack-aware optimizations? And would it help if KafkaStreamsAssignment had some methods for mutating the set of assigned tasks so that we can use the KafkaStreamsAssignment class directly, instead of always translating between maps of KafkaStreamsAssignment and maps of Set<AssignedTask>? I was thinking about this earlier since it also comes up a lot in the StickyTaskAssignor

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:104]: @apourchet FYI I just added these javadocs to the KIP so you can copy/paste them in

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 9]
```diff
@@ -74,20 +198,219 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
         final ApplicationState applicationState,
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments
     ) {
+        if (!hasValidRackInformation(applicationState)) {
+            throw new IllegalStateException("Cannot perform rack-aware assignment optimizations with invalid rack information.");
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: ditto here (update to warning log instead of exception)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 10]
```diff
@@ -16,14 +16,63 @@
...
+     * Return a "no-op" assignment that just copies the previous assignment of tasks to KafkaStreams clients
+     *
+     * @param applicationState the metadata and other info describing the current application state
+     *
+     * @return a new map containing an assignment that replicates exactly the previous assignment reported
+     *         in the applicationState
+     */
+    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
+        final ApplicationState applicationState
+    ) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Kafka Streams formatting: first parameter always on the same line (no matter how ridiculously long it gets)

```suggestion
    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(final ApplicationState applicationState) {
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 4 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:104]: I think we're still missing the javadocs?

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 11]
```diff
@@ -74,20 +198,220 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
...
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    public static boolean hasValidRackInformation(final ApplicationState applicationState) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: needs to be private, otherwise it would be a public API which we don't want (and we don't need IIUC)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 12]
```diff
@@ -16,14 +16,63 @@
...
+            });
+
+            final KafkaStreamsAssignment newAssignment = KafkaStreamsAssignment.of(processId, tasks);
+            assignments.put(processId, newAssignment);
+        });
+        return assignments;
+    }
+
     /**
      * Assign standby tasks to KafkaStreams clients according to the default logic.
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Let's also just go ahead and remove this utility method for now while we're at it, so we don't have to remember to remove it later if we don't implement it before the 3.8 code freeze

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 13]
```diff
@@ -16,78 +16,408 @@
...
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    private static boolean hasValidRackInformation(final ApplicationState applicationState) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: We should check the rackAwareAssignmentStrategy at the top of this method (also maybe rename this method to something like `canPerformRackAwareOptimization` to account for that)

```
final String rackAwareAssignmentStrategy = applicationState.assignmentConfigs.rackAwareAssignmentStrategy();
if (StreamsConfig.RACK_AWARE_ASSIGNMENT_STRATEGY_NONE.equals(rackAwareAssignmentStrategy)) {
    return false;
}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 14]
```diff
@@ -16,78 +16,408 @@
...
+        }
+        return cost;
+    }
+
+    /**
+     * This function returns whether the current application state has the required rack information
+     * to make assignment decisions with.
+     * This includes ensuring that every client has a known rack id, and that every topic partition for
+     * every logical task that needs to be assigned also has known rack ids.
+     * If a logical task has no source topic partitions, it will return false.
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: This comment should be updated as well. I think it's ok to keep it simple (or even remove these javadocs altogether) instead of trying to hit all the cases in the javadocs, since that will (clearly) be easy to forget to update and quickly go out of date. This is another one of those "the code should be its own documentation" cases -- docs and comments are never up to date in Streams, too many people changing things without knowing where comments are that reference something they've changed, hence it's best to just use clear method names and easily readable code rather than try to explain what something does in docs/comments. 

Just the general code style/philosophy of Kafka/Kafka Streams 🤷‍♀️

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 15]
```diff
@@ -16,78 +16,408 @@
...
+            if (!hasValidRackInformation(task)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final KafkaStreamsState state) {
+        if (!state.rackId().isPresent()) {
+            LOG.error("Client " + state.processId() + " doesn't have rack configured.");
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
            LOG.error("KafkaStreams client {} doesn't have a rack id configured.", state.processId().id());
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 16]
```diff
@@ -16,78 +16,408 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final TaskInfo task) {
+        for (final TaskTopicPartition topicPartition : task.topicPartitions()) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Just comparing with the old `#canEnableRackAwareAssignor` code, we only require the source topic partitions to have rack ids for the active task optimization, and only need the changelog partitions to have rack ids for the standby task optimization. So we should probably add a boolean parameter to indicate whether w'ere calling from the optimizeActive or optimizeStandby method, and then only need to validate rack ids for partitions of the corresponding type

edit: we actually need rack ids for both partitions types for the active case, it's just for the standby case that we only care about the changelogs specifically

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 17]
```diff
@@ -16,78 +16,408 @@
...
+     * This function returns a copy of the old collection of {@code KafkaStreamsAssignment} with tasks
+     * moved according to the {@param reassigned} tasks and {@param unassigned} tasks.
+     *
+     * @param kafkaStreamsAssignments the collection to start from when moving tasks from process to process
+     * @param reassigned              the map from process id to tasks that this client has newly been assigned
+     * @param unassigned              the map from process id to tasks that this client has newly been unassigned
+     *
+     * @return the new mapping from processId to {@code KafkaStreamsAssignment}.
+     */
+    private static Map<ProcessId, KafkaStreamsAssignment> processTaskMoves(final Collection<KafkaStreamsAssignment> kafkaStreamsAssignments,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: I can't remember if I've brought this up already but I've noticed we have to do a lot of migrating between different data sets and converting types seemingly all because a KafkaStreamsAssignment is essentially immutable. You could do this in a separate PR so it wouldn't block anything, but WDYT about removing the unmodifiableSet from the `Set<AssignedTask> assignment` field of KafkaStreamsAssignment?

Or, probably better API design, offer some methods for adding/removing tasks rather than exposing the set itself. ie `KafkaStreamsAssignment.addAssignedTask(AssignedTask newTask)` and `KafkaStreamsAssignment.removeAssignedTask(AssignedTask removedTask)`? I'm thinking that would let us simplify a lot of things, here and in the actual assignor implementations. I also wonder if we're going to want that when we implement our own custom assignors, if we just want to adjust an assignment slightly so that we aren't forced to rebuild things again and again. 

Thoughts? We can add the new APIs in a separate and small PR so it shouldn't get in the way of anything else. Then we can migrate existing code on our own time and use it for new code as needed

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 18]
```diff
@@ -16,78 +16,408 @@
...
+        final Map<ProcessId, KafkaStreamsAssignment> optimizedResult = new HashMap<>();
+        for (final KafkaStreamsAssignment oldAssignment : kafkaStreamsAssignments) {
+            final ProcessId processId = oldAssignment.processId();
+            final Set<AssignedTask> newTasks = reassigned.getOrDefault(processId.id(), new HashSet<>());
+            final Set<TaskId> unassignedTasksForProcess = unassigned.getOrDefault(processId.id(), new HashSet<>());
+            for (final AssignedTask previouslyAssignedTask : oldAssignment.assignment()) {
+                if (unassignedTasksForProcess.contains(previouslyAssignedTask.id())) {
+                    continue;
+                }
+                newTasks.add(previouslyAssignedTask);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
                if (!unassignedTasksForProcess.contains(previouslyAssignedTask.id())) {
                    newTasks.add(previouslyAssignedTask);
                }
```

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Sorry to keep harping on this, I actually used to use this kind of exit-early style all the time and had one of the older committers always yelling at me to keep things consolidated. It was annoying at the time but now I agree with them completely, it's easier to read if the logic all flows in one direction without early `return` or `continue` breaking it up

(is this just the Kafka Streams version of "goto considered harmful"? 🤔  lol)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 19]
```diff
@@ -16,78 +16,408 @@
...
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
     ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+        if (tasks.isEmpty()) {
+            return kafkaStreamsAssignments;
+        }
+
+        if (!hasValidRackInformation(applicationState)) {
+            LOG.warn("Cannot optimize active tasks with invalid rack information.");
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: meh, I'd remove the warning here. We log whatever the actual issue is when we hit it inside the `#hasValidRackInformation` call, which is more useful since we know the specific reason in that context. 

(Same for the optimize standby method)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 20]
```diff
@@ -16,78 +16,408 @@
...
+    }
+
     /**
-     * Return a "no-op" assignment that just copies the previous assignment of tasks to KafkaStreams clients
      *
-     * @param applicationState the metadata and other info describing the current application state
+     * @return the traffic cost of assigning this {@param task} to the client {@param streamsState}.
+     */
+    private static int getCrossRackTrafficCost(final Set<TaskTopicPartition> topicPartitions,
+                                               final Optional<String> clientRack,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Seems silly to have the input parameter type be Optional since that seems to suggest this parameter can be empty. Yet we actually throw an exception if it ever is. We should make this a `String` and move the check to wherever we're invoking this method (or even skip the `#isPresent` check altogether, see my comment below)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 21]
```diff
@@ -16,78 +16,408 @@
...
+                                               final int crossRackTrafficCost) {
+        if (!clientRack.isPresent()) {
+            throw new IllegalStateException("Client doesn't have rack configured.");
+        }
+
+        int cost = 0;
+        for (final TaskTopicPartition topicPartition : topicPartitions) {
+            final Optional<Set<String>> topicPartitionRacks = topicPartition.rackIds();
+            if (topicPartitionRacks == null || !topicPartitionRacks.isPresent()) {
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: No reason for a null check here imo, remember we should have validated this/should be able to trust the assignment inputs. On that note, do we even need to do the `#isPresent` check? I mean didn't we make sure all these are filled in during the `hasValidRackInformation` call at the top? 

That said it doesn't hurt, and I know the IDE will probably show a warning if you don't include the `#isPresent` check, since I'm guessing it's not smart enough to know that we checked this all the way at the start. So we can leave in the `#isPresent` check if you want. Your call on this

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 22]
```diff
@@ -16,78 +16,408 @@
...
+                throw new IllegalStateException("TopicPartition " + topicPartition + " has no rack information.");
+            }
+
+            if (topicPartitionRacks.get().contains(clientRack.get())) {
+                continue;
+            }
+
+            cost += crossRackTrafficCost;
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]:
```suggestion
            if (!topicPartitionRacks.get().contains(clientRack.get())) {
                cost += crossRackTrafficCost;;
            }
```

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 23]
```diff
@@ -16,78 +16,408 @@
...
+                return previousTaskIdsByProcess.get(processId).contains(taskId);
+            },
+            (taskId, processId, inCurrentAssignment, unused0, unused1, unused2) -> {
+                final int assignmentChangeCost = !inCurrentAssignment ? nonOverlapCost : 0;
+                final Optional<String> clientRack = clientRacks.get(processId);
+                final Set<TaskTopicPartition> topicPartitions = topicPartitionsByTaskId.get(taskId).stream().filter(tp -> {
+                    return isStandby ? tp.isChangelog() : true;
+                }).collect(Collectors.toSet());
+                return assignmentChangeCost + getCrossRackTrafficCost(topicPartitions, clientRack, crossRackTrafficCost);
+            },
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: this is kind of nit-y, so feel free to ignore, but it's really hard to follow these complex inline implementations for the function pointers. Compare with the RackAwareTaskAssignor implementation, where it just passes in a simple `this##getCost` and you can find all the logic in one place. 

For a concrete example, I had left a comment about missing the += nonOverlapCost part at the end of the cost function, but just happened to notice that's actually accounted for by the `assignmentChangeCost` that's inlined right here. It would be easier to follow if everything in this cost function was all wrapped up in one method, especially when there are multiple function pointer parameters inlined right on top of each other making it hard to tell what code belongs to which function pointer 😵

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 24]
```diff
@@ -16,78 +16,408 @@
...
+        final Map<TaskId, UUID> clientByTask = new HashMap<>();
+        final Map<UUID, Integer> taskCountByClient = new HashMap<>();
+        final Graph<Integer> graph = graphConstructor.constructTaskGraph(
+            clientIds,
+            taskIds,
+            clientsUuidByUuid,
+            clientByTask,
+            taskCountByClient,
+            (processId, taskId) -> {
+                return previousTaskIdsByProcess.get(processId).contains(taskId);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Unfortunately this is still incorrect -- the function pointer in this parameter is supposed to return whether the task is assigned to this client in the current iteration of the assignment. You'll notice in the RackAwareTaskAssignor that this essentially boils down to checking `ClientState#hasAssignedTask`, which corresponds to what is _currently_ assigned to that ClientState, not what was _previously_ assigned to that clientState.

The main fix will be in what we pass in as the `Map<UUID, Set<TaskId>>`, which I'll try to figure out and address in another comment, but for this method we should at least change the name from `previousTaskIdsByProcess` to `assignedTasksForProcess` or something like that.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 25]
```diff
@@ -16,78 +16,408 @@
...
+            crossRackTrafficCost,
+            nonOverlapCost,
+            true,
+            true
+        );
+        LOG.info("Assignment before standby task optimization has cost {}", currentCost);
         throw new UnsupportedOperationException("Not Implemented.");
     }
 
+    private static long computeTaskCost(final Set<TaskInfo> tasks,
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: This is literally only used to get the cost at the beginning of the method, right? (and only for logging purposes it seems? Weird but let's keep the original logic as usual) I know it's like this in the original implementation but as a small improvement, can we rename this method to `#computeInitialCost` or something like that which makes its purpose more clear?

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 26]
```diff
@@ -16,78 +16,408 @@
...
+            nonOverlapCost,
+            hasReplica,
+            isStandby
+        );
+        return new AssignmentGraph(graph, clientByTask, taskCountByClient);
+    }
+
+    /**
+     * This internal structure is used to keep track of the graph solving outputs alongside the graph
+     * structure itself.
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:315]: This is a great abstraction, I wish the original graph code had this container class for outputs (or like any documentation whatsoever lol) because I was really struggling to understand what was an input and what was an output for which method.

 Would be nice to clean up at some point (we can just file an AK Jira ticket for post-KIP-924 cleanup work that isn't strictly necessary, maybe someone else will pick it up)

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 27]
```diff
@@ -16,78 +16,408 @@
...
+                Collectors.toMap(state -> state.processId().id(), KafkaStreamsState::rackId));
+        final Map<UUID, Set<TaskId>> previousTaskIdsByProcess = kafkaStreamsStates.values().stream().collect(Collectors.toMap(
+            state -> state.processId().id(),
+            KafkaStreamsState::previousActiveTasks
+        ));
+        final Map<TaskId, Set<TaskTopicPartition>> topicPartitionsByTaskId = applicationState.allTasks().stream()
+            .filter(taskInfo -> tasks.contains(taskInfo.id()))
+            .collect(Collectors.toMap(TaskInfo::id, TaskInfo::topicPartitions));
+
+        final List<TaskId> taskIds = new ArrayList<>(tasks);
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: We're computing some things over and over again, for example we reconstruct this same `List<TaskId>` three times for the active task optimization. The `List<UUID` gets computed twice. These may be smallish sets but the assignor should still be sensitive to memory consumption and performance. And just from a review perspective, I'm getting a bit overwhelmed by how many/how often we're creating new variations of these data structures via `.stream()`  😅 

It seems like most of the rack aware/graph methods have the same input parameters, so we may as well just construct them all at the top and then pass those around as needed.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 28]
```diff
@@ -16,78 +16,408 @@
...
+
+        final Map<UUID, Optional<String>> clientRacks = clients.values().stream().collect(
+            Collectors.toMap(state -> state.processId().id(), KafkaStreamsState::rackId));
+
+        final Map<UUID, Set<TaskId>> taskIdsByProcess = clients.values().stream().collect(
+            Collectors.toMap(state -> state.processId().id(), state -> {
+                if (isStandby) {
+                    return state.previousStandbyTasks();
+                }
+                return state.previousActiveTasks();
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: this is what we're passing in that ultimately becomes the `hasAssignedTask` predicate for the graph constructors, right? We actually need the current assignment (so what's in the KafkaStreamsAssignment) not the previous assignment (ie what's in the KafkaStreamsState). See https://github.com/apache/kafka/pull/16033/files#r1618135137

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: I just realized that the wording of my comment here was potentially misleading. Sorry about that. What I meant by this

> However both callers of this method construct the tasksIdsByProcess map based on the previous task assignment, not the current one.

was not that the variable name should be updated to `previousTaskIdsByProcess`, but rather the contents of the map passed in by the callers should be updated to pass in the current assignment, not the previous assignment. Sorry for all the confusion around this.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 28 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Ah I just reread my [original comment](https://github.com/apache/kafka/pull/16033#discussion_r1618282249) about this and I think it's my fault for poorly wording what I was trying to actually suggest there. Anyways this should all be cleared up as part of the [POC](https://github.com/apourchet/kafka/pull/1) so hopefully we don't have to think about this ever again 😄

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 24 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Addressed by https://github.com/apourchet/kafka/pull/1

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 27 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: I ended up having to redo a bunch of the data structures as part of the KafkaStreamsAssignment POC ([this](https://github.com/apourchet/kafka/pull/1)) so I happened to address this comment as well. 

You can see in the POC PR -- there's not actually very much we need to compute if we just do it all up front

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 17 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: Alright I went ahead and tried this out for myself, I think it helps a lot to simplify things in this PR so it makes sense to incorporate here rather than doing it as an entirely separate PR. Lmk your thoughts as always: https://github.com/apourchet/kafka/pull/1

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java (hunk 23 above)
Comment (apourchet) [reply streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java]: We can't really have a simple this##getCost function because we have no `this`, which means we cannot easily move around the `clientRacks`, or `topicPartitionsByTaskId`. It's going to have to be a function pointer no matter how we spin it unfortunately.

File: streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java [hunk 29]
```diff
@@ -16,78 +16,395 @@
...
+     * @param tasks                   the set of tasks to reassign if possible. Must already be assigned to a KafkaStreams client
      *
-     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default
-     *         rack-aware assignment for active tasks
+     * @return a map with the KafkaStreamsAssignments updated to minimize cross-rack traffic for active tasks
      */
     public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTasks(
         final ApplicationState applicationState,
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java:104]: need to fix the indentation (can be done in a followup PR)

**Review by ableegoldman — APPROVED** [review]
LGTM, with one minor formatting issue that can be fixed in a followup PR

**Comment by ableegoldman** [comment]
Test failures are unrelated, merging to trunk
